File: toolkit/NLA/faspsolver/tutorial/main/poisson-amg.c
/*! \file  poisson-amg.c
 *
 *  \brief The first test example for FASP: using AMG to solve 
 *         the discrete Poisson equation from P1 finite element.
 *         C version.
 *
 *  \note  Solving the Poisson equation (P1 FEM) with AMG: C version
 *
 *---------------------------------------------------------------------------------
 *  Copyright (C) 2011--Present by the FASP team. All rights reserved.
 *  Released under the terms of the GNU Lesser General Public License 3.0 or later.
 *---------------------------------------------------------------------------------
 */

#include "fasp.h"
#include "fasp_functs.h"

/**
 * \fn int main (int argc, const char * argv[])
 *
 * \brief This is the main function for the first example.
 *
 * \author Chensong Zhang
 * \date   12/21/2011
 * 
 * Modified by Chensong Zhang on 09/22/2012
 */
int main (int argc, const char * argv[]) 
{
    input_param     inparam;  // parameters from input files
    AMG_param       amgparam; // parameters for AMG
    
    printf("\n========================================");
    printf("\n||   FASP: AMG example -- C version   ||");
    printf("\n========================================\n\n");

    // Step 0. Set parameters: We can use ini/amg.dat
    fasp_param_set(argc, argv, &inparam);
    fasp_param_init(&inparam, NULL, &amgparam, NULL, NULL);
    
    // Set local parameters using the input values
    const int print_level = inparam.print_level;
    
    // Step 1. Get stiffness matrix and right-hand side
    // Read A and b -- P1 FE discretization for Poisson. The location
    // of the data files is given in "ini/amg.dat".
    dCSRmat A;
    dvector b, x;
    char filename1[512], *datafile1;
    char filename2[512], *datafile2;
    
    // Read the stiffness matrix from matFE.dat
    memcpy(filename1, inparam.workdir, STRLEN);
    datafile1="csrmat_FE.dat"; strcat(filename1, datafile1);
    
    // Read the RHS from rhsFE.dat
    memcpy(filename2, inparam.workdir, STRLEN);
    datafile2="rhs_FE.dat"; strcat(filename2, datafile2);
    
    fasp_dcsrvec_read2(filename1, filename2, &A, &b);
    
    // Step 2. Print problem size and AMG parameters
    if (print_level>PRINT_NONE) {
        printf("A: m = %d, n = %d, nnz = %d\n", A.row, A.col, A.nnz);
        printf("b: n = %d\n", b.row);
        fasp_param_amg_print(&amgparam);
    }
    
    // Step 3. Solve the system with AMG as an iterative solver
    // Set the initial guess to be zero and then solve it
    // with AMG method as an iterative procedure
    fasp_dvec_alloc(A.row, &x);
    fasp_dvec_set(A.row, &x, 0.0);
    
    fasp_solver_amg(&A, &b, &x, &amgparam);
    
    // Step 4. Clean up memory
    fasp_dcsr_free(&A);
    fasp_dvec_free(&b);
    fasp_dvec_free(&x);
    
    return FASP_SUCCESS;
}

/*---------------------------------*/
/*--        End of File          --*/
/*---------------------------------*/

File: toolkit/NLA/faspsolver/tutorial/main/poisson-amg.f90
!> \file  poisson-amg.f90
!>
!> \brief The first test example for FASP: using AMG to solve
!>        the discrete Poisson equation from P1 finite element.
!>        F90 version.
!>
!> \note  Solving the Poisson equation (P1 FEM) with AMG: F90 version
!>
!>---------------------------------------------------------------------------------
!>  Copyright (C) 2011--Present by the FASP team. All rights reserved.
!>  Released under the terms of the GNU Lesser General Public License 3.0 or later.
!>---------------------------------------------------------------------------------

program test

  implicit none

  double precision, dimension(:), allocatable :: u, b, a
  integer,          dimension(:), allocatable :: ia, ja

  integer          :: iufile, n, nnz, i, prt_lvl, maxit
  double precision :: tol

  print*, ""
  write(*,"(A)") "=========================================="
  write(*,"(A)") "||   FASP: AMG example -- F90 version   ||"
  write(*,"(A)") "=========================================="
  print*, ""

  ! Step 0: user defined variables
  prt_lvl = 3
  maxit = 100
  tol = 1.0d-6
  iufile = 1

  ! Step 1: read A and b 

  !===> Read data A from file
  open(unit=iufile,file='../data/csrmat_FE.dat')

  read(iufile,*) n
  allocate(ia(1:n+1))   
  read(iufile,*) (ia(i),i=1,n+1)   

  nnz=ia(n+1)-ia(1)
  allocate(ja(1:nnz),a(1:nnz))
  read(iufile,*) (ja(i),i=1,nnz)
  read(iufile,*) (a(i),i=1,nnz)   

  close(iufile)

  !===> Read data b from file
  open(unit=iufile,file='../data/rhs_FE.dat')

  read(iufile,*) n
  allocate(b(1:n))
  read(iufile,*) (b(i),i=1,n)

  close(iufile)

  !===> Shift the index to start from 0 (for C routines)
  forall (i=1:n+1) ia(i)=ia(i)-1
  forall (i=1:nnz) ja(i)=ja(i)-1

  ! Step 2: Solve the system 

  !===> Initial guess
  allocate(u(1:n))
  u=0.0d0 
  call fasp_fwrapper_dcsr_amg(n,nnz,ia,ja,a,b,u,tol,maxit,prt_lvl)
 
  ! Step 3: Clean up memory
  deallocate(ia,ja,a)
  deallocate(b,u)

end program test

!/*---------------------------------*/
!/*--        End of File          --*/
!/*---------------------------------*/
